src/lib/arch/riscv/Architecture.cc

```diff
@@ -184,6 +198,17 @@ uint8_t Architecture::predecode(const void* ptr, uint8_t bytesAvailable,
     bool success =
         cs_disasm_iter(capstoneHandle, &encoding, &size, &address, &rawInsn);
 
+    if (!success || rawInsn.opcode == RISCV_INS_INVALID) {
+      // Invalid decoding, potentially read over the end of the valid buffer.
+      // BAIL. Do not add to cache as may be incorrect data
+      return 0;
+    }
+
+    if (rawInsn.size > bytesAvailable) {
+      // Too many bytes read. BAIL
+      return 0;
+    }
```

Comment:
See above comment RE this clause. Interestingly I haven't had that assert trigger probably because it is within the loop buffer logic. I needed to comment out the other assertion in the fetch unit for this reason. I have added my own comment to this PR for discussion on the correct behaviour of an early bail, see [here](https://github.com/UoB-HPC/SimEng/pull/368#discussion_r1448990715)

Reply:
This assert is not within the loop buffer logic? It comes right after the two if clauses which deal with the loop buffer, inside the for loop which predecodes instructions and places them into the output buffer

Reply:
Implementation has changed. Outdated, resolving